Reactive value generation

DavideD opened the pull request:
Fixes #1099

gavinking commented:
OK, this patch is bigger than expected, so can you give me until tomorrow evening to look it over?

DavideD commented:
Sure.

On Tue, 21 Dec 2021, 21:51 Gavin King, ***@***.***> wrote:

> OK, this patch is bigger than expected, so can you give me until tomorrow
> evening to look it over?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hibernate/hibernate-reactive/pull/1103#issuecomment-999085968>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEIQ5LZSFLRC3KHBUN6I33USDSGBANCNFSM5KQDIFMQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

gavinking commented on hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java:
```diff
@@ -201,102 +211,119 @@ default ReactiveConnection getReactiveConnection(SharedSessionContractImplemento
+			for ( int i = 0; i < strategies.length; i++ ) {
+				final int index = i;
+				final InMemoryValueGenerationStrategy strategy = strategies[i];
+				if ( strategy != null && strategy.getGenerationTiming().includesInsert() ) {
```
What happens to in-memory (nonreactive) value generators?

DavideD replied:
Not sure I understand the question. But I've just moved the existing code

gavinking commented on hibernate-reactive-core/src/main/java/org/hibernate/reactive/tuple/MutinyValueGenerator.java:
```diff
@@ -0,0 +1,31 @@
+	@Override
+	default T generateValue(Session session, Object owner) {
+		throw new UnsupportedOperationException( "Use generateValue(Mutiny.Session, Object) instead" );
+	}
```
Ugh. This isn't very nice.

gavinking commented:
So I think we need to find some way of removing the inheritance from `ValueGeneration`.

That's a problem I had not noticed.

DavideD commented:
Yeah, I agree, but I cannot think of a better solution right now.

gavinking commented:
Right, neither can I.

DavideD commented:
Maybe having abstract classes instead of interfaces is slightly better in this scenario: 2998ff0f5f3ff82d1d787fbb836a5ef8b27c4442

gavinking commented:
Because it lets you redeclare the method `final`, you mean?

DavideD commented:
Yes

DavideD commented:
I'm also going to add the `@Incubating` annotation to clarify that's a temporary solution.
This way people people have at least a workaround for this feature.

gavinking commented:
> I'm also going to add the `@Incubating` annotation to clarify that's a temporary solution.
> This way people have at least a workaround for this feature.

That's a good idea.

DavideD commented:
Done. I'm going to merge it as it is if there is no strong opposition

gavinking commented:
My only question was: does it break in-memory impls of `ValueGeneration`, or do they still work?

DavideD commented:
How can I test it? All our tests are passing ...

gavinking commented:
Well, the `@CurrentUser` annotation is our test for that.

gavinking commented:
Sorry, I mean `@GeneratorType(type = LoggedUserGenerator.class, when = GenerationTime.ALWAYS)`

DavideD commented:
I've moved the code around but I don't think I've changed anything of the original logic (not on purpose at least).
The core of the change is the method [ReactiveAbstractEntityPersister#generateValue](https://github.com/DavideD/hibernate-reactive/blob/b235ae9e3071c6bbd41a6046cab84fc53d1da622/hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java#L753).

If a user will use the new reactive classes, Hibernate Reactive will call the reactive methods. Implementations of `ValueGenerator` [will still work](https://github.com/DavideD/hibernate-reactive/blob/b235ae9e3071c6bbd41a6046cab84fc53d1da622/hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java#L772).

We test the new classes via `GeneratedProperty*Test` but we also have a test for `ValueGenerator` via `TimestampTest` because it tests the use of the annotations `@CreationTimestamp` and `@UpdateTimestamp` (they use a valuegenerator underneath). Maybe adding a direct test for ValueGenerator would be wise though.

The drawback of allowing an implementation of `ValueGenerator` is that a user can access the ReactiveSession, something that normally would be hidden.

Overall I think our tests cover most cases (I was pleasantly surprised when I've seen the test suite fails when I've started to work on this) but  if you have some other test in mind, let me know.

DavideD commented:
Merged.
This should be good enough for now.
